init.lua

```diff
@@ -283,55 +332,27 @@ minetest.register_node("anvil:anvil", {
 		return stack:get_count()
 	end,
 
-	on_rightclick = function(pos, node, clicker, itemstack)
-		if not clicker or not itemstack then
-			return
+	on_rotate = function(pos, node, user, mode, new_param2)
+		if mode == screwdriver.ROTATE_FACE then
+			return 1
 		end
-		local meta = minetest.get_meta(pos)
-		local name = clicker:get_player_name()
-		local owner = meta:get_string("owner")
-		local shared = meta:get_int("shared") == 1
 
-		if name ~= owner and not shared then
-			return itemstack
-		end
-		if itemstack:get_count() == 0 then
-			local inv = meta:get_inventory()
-			if not inv:is_empty("input") then
-				local return_stack = inv:get_stack("input", 1)
-				inv:set_stack("input", 1, nil)
-				local wield_index = clicker:get_wield_index()
-				clicker:get_inventory():set_stack("main", wield_index, return_stack)
-				if shared then
-					meta:set_string("infotext", S("Shared anvil"))
-				else
-					meta:set_string("infotext", S("@1's anvil", owner))
-				end
-				remove_item(pos, node)
-				return return_stack
-			end
-		end
-		local this_def = minetest.registered_nodes[node.name]
-		if this_def.allow_metadata_inventory_put(pos, "input", 1, itemstack:peek_item(), clicker) > 0 then
-			local s = itemstack:take_item()
-			local meta = minetest.get_meta(pos)
-			local inv = meta:get_inventory()
-			inv:add_item("input", s)
-			local meta_description = s:get_meta():get_string("description")
-			if "" ~= meta_description then
-				if shared then
-					meta:set_string("infotext", S("Shared anvil"))
-				else
-					meta:set_string("infotext", S("@1's anvil", owner) .. "\n" .. meta_description)
-				end
-			end
-			meta:set_int("informed", 0)
-			update_item(pos, node)
+		local wielded = user and user:get_wielded_item()
+
+		if wielded:get_name() ~= "screwdriver:screwdriver" then
+			return 1
 		end
 
-		return itemstack
+		core.after(0,function()
+			local z=anvil_rightclick(pos,node,user,wielded)
+			user:set_wielded_item(z)
+		end)
```

Comment:
this is unreliable, as the wielded index can change between server steps.

Reply:
> this is unreliable, as the wielded index can change between server steps.
maybe I can save the wielded index into a local variable and use that in the lambda. Would that be acceptable for you?

Reply:
sorry i haven't responded. the inventory itself can change between server steps, you may still be clobbering an item that you don't intend to. what problem are you trying to solve by not calling `anvil_rightclick` immediately?

Reply:
well, to be precise as per my test right now, the screwdriver gets duplicated: if I rightclick without core.after a screwdriver appears on my anvil, while it still remains in my hand

Reply:
maybe we need to change the api and make sure that whenever the on_rotate returns with an ItemStack (and not with true-false-0-1-2-3) the wielded item gets replaced with that ItemStack by the caller? Or maybe make this tunction return with a pair of "whatever used to be returned until now",optional ItemStack. That would be a solution... What do you think? I might be even able to make that PR. I suppose noone tried to return from an on_rotate with an ItemStack, so both of those solutions would be backwards compatible.

Reply:
i wish the screwdriver API was a bit better, but since MTG is feature frozen, there's probably no chance of that. the way i solve this issue in server integration code is to override the screwdriver's `on_place` behavior, so that if that is called while pointing at an anvil (or e.g. an item frame), the node's `on_rightclick` callback ends up being invoked instead ([an example](https://gitea.your-land.de/your-land/yl_commons/src/branch/yl_stable/features/tool_rightclick_behavior.lua)). 

but let's see if we can get this approach to work.

you are right, of course you can't call `anvil_rightclick` directly inside of the `on_rotate` callback immediately, because that `on_rotate` call is handled inside the screwdriver's `on_place` callback, which returns an itemstack which overrides the player's wielded item. 

the `after` callback is an interesting approach, but the world can change a lot between server steps. the following things can all happen:
* the player selects a different item in their inventory
* the player uses and breaks the screwdriver
* the screwdriver is modified by another mod (e.g. perhaps it is repaired)
* the player's inventory is modified by another mod entirely in an unpredictable way
* the player disconnects or logs out
* the *anvil node* is removed or replaced somehow

the solution is to move a lot of the logic *inside* the `after` callback, to check that the world is still the way we expect it. if it is, *then* we can try to place the screwdriver on the anvil. 

this is what i've come up with:

```lua
	on_rotate = function(pos, node, user, mode, new_param2)
		if mode == screwdriver.ROTATE_FACE then
			return
		end

		if not minetest.is_player(user) then
			return
		end

		local player_name = user:get_player_name()
		local wield_list = user:get_wield_list()
		local wield_index = user:get_wield_index()

		minetest.after(0,function()
			local player = minetest.get_player_by_name(player_name)
			if not player then
				return
			end

			local inv = player:get_inventory()
			local wielded = inv:get_stack(wield_list, wield_index)

			if wielded:get_name() ~= "screwdriver:screwdriver" then
				return
			end

			local node_now = minetest.get_node(pos)
			if node_now.name ~= "anvil:anvil" then
				return
			end

			local z = anvil_rightclick(pos, node_now, player, wielded)
			inv:set_stack(wield_list, wield_index, z)
		end)

		return false
	end,
```

thoughts?

Reply:
hello I'll revert back to you with my PR updated once I find the time to inspect and try your code. thank you for your patience.

Reply:
ok I made one minor change, I renamed z to tool_after_rightclicking